src/js/links.js

```diff
@@ -0,0 +1,195 @@
+ * }
+ */
+ /*globals gPinnedLinks, gBlockedLinks, ProviderManager, DirectoryLinksProvider*/
+ /*jshint worker:true*/
+
+"use strict";
+(function(exports) {
+  const Links = {
+
+    maxNumLinks: 100,
+
+    /**
+     * A mapping from each provider to an object { sortedLinks, siteMap, linkMap }.
+     * sortedLinks is the cached, sorted array of links for the provider.
+     * siteMap is a mapping from base domains to URL count associated with the domain.
+     *         The count does not include blocked URLs. siteMap is used to look up a
+     *         user's top sites that can be targeted with a suggested tile.
+     * linkMap is a Map from link URLs to link objects.
+     */
+    _providers: new Map(),
+
+    /**
+     * A list of objects that are observing links updates.
+     */
+    _observers: [],
+
+    /**
+     * Compares two links.
+     *
+     * @param {Link} aLink1 The first link.
+     * @param {Link} aLink2 The second link.
+     * @return {Integer} -1 if aLink1 is ordered before aLink2, 0 if
+     *         aLink1 and aLink2 have the same ordering, or 1 if
+     *         aLink1 is ordered after aLink2.
+     *
+     * NOTE: compareLinks's this object is bound to Links below.
+     */
+    compareLinks(aLink1, aLink2) {
+      // The properties of link objects used to sort them.
+      const requiredProps = ["frecency", "lastVisitDate", "url"];
+      if (!requiredProps.every(prop => prop in aLink1 && prop in aLink2)) {
+        throw new Error("Comparable link missing required property");
+      }
+      let result = aLink2.frecency - aLink1.frecency ||
+                   aLink2.lastVisitDate - aLink1.lastVisitDate ||
+                   aLink1.url.localeCompare(aLink2.url);
+      return result === 0 ? 0 : result / Math.abs(result);
+    },
+
+    /**
+     * Merges the cached lists of links from all providers whose lists are cached.
+     *
+     * @return {Array} The merged list.
+     */
+    _getMergedProviderLinks() {
+      return Array.from(this._providers.values())
+        .filter(provider => provider.sortedLinks)
+        .reduce((arr, provider) => arr.concat(provider.sortedLinks.slice()), [])
+        .sort(this.compareLinks)
+        .slice(0, this.maxNumLinks);
+    },
+
+    /**
+     * Gets the current set of links contained in the grid.
+     *
+     * @return {Object} The links in the grid.
+     */
+    getLinks() {
+      let pinnedLinks = gPinnedLinks.links.slice();
+      let links = this._getMergedProviderLinks();
+
+      let sites = pinnedLinks
+        .filter(link => link)
+        .map(link => ProviderManager.extractSite(link.url))
+        .reduce((set, site) => set.add(site), new Set());
+
+      // Filter blocked and pinned links and duplicate base domains.
+      links = links.filter(function(link) {
+        let site = ProviderManager.extractSite(link.url);
+        if (!site || sites.has(site)) {
+          return false;
+        }
+        sites.add(site);
```

Comment:
Please avoid side effects (i.e., calling `.add()`). Only add sites after filtering - see previous comments around this code.